Refactoring AutomationId at the accessible object classes #6603

dkazennov opened the pull request:
Fixes #6603
This PR finishes https://github.com/dotnet/winforms/pull/6641 (adds unit tests)

## Proposed changes

- The AutomationId property is moved to the AccessibleObject base class
- This property returns null at AccessibleObject
- It is returned at the GetPropertyValue method of AccessibleObject for the case AutomationIdPropertyId
- The current implementations in the 4 classes derived from AO are kept the same way, but they are marked as overrides
- Returning of the Owner.Name is moved to the ControlAccessibleObjectObject class to make it universal for all its descendants
- All duplicate code with the same logic is cleaned out in the derived classes
- Initial discussion on this topic is here: https://github.com/dotnet/winforms/pull/6429#issuecomment-1009286397

Unit tests:

1) AccessibleObject - AccessibleObject_GetPropertyValue_ReturnsNull (two unit tests combined)
2) DataGridViewCellAccessibleObject - DataGridViewCellAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected (new)
3) ListViewItemBaseAccessibleObject - ListViewItemBaseAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected (new)
4) ListViewSubItemAccessibleObject - ListViewSubItemAccessibleObject_GetPropertyValue_returns_correct_values (existing)
5) ControlAccessibleObject - ControlAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected (new)
6) ButtonBaseAccessibleObject - ButtonBaseAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected (new)
7) CheckBoxAccessibleObject - CheckBoxAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected (existing)
8) GroupBoxAccessibleObject - GroupBoxAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected (new)
9) LabelAccessibleObject - LabelAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected (two unit tests combined)
10) ListViewAccessibleObject - ListViewAccessibleObject_GetPropertyValue_returns_correct_values (existing)
11) PanelAccessibleObject - PanelAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected (existing)
12) PictureBoxAccessibleObject - PictureBoxAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected (existing)
13) RadioButtonAccessibleObject - RadioButtonAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected (existing)
14) ScrollBarAccessibleObject - ScrollBarAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected (existing)
15) SplitterAccessibleObject - SplitterAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected (existing)
16) TabControlAccessibleObject - TabControlAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected (two unit tests combined)
17) TabPageAccessibleObject - TabPageAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected (two unit tests combined)
18) TrackBarAccessibleObject - TrackBarAccessibilityObject_GetPropertyValue_Invoke_ReturnsExpected (two unit tests combined)

Combined existed NamePropertyId tests and new AutomationIdPropertyId tests:
{
		TrackBarAccessibilityObject_GetPropertyValue_Invoke_ReturnsExpected
		TabPageAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected
		TabControlAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected
		LabelAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected
}
just like existing:
{
		CheckBoxAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected
		PanelAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected
		PictureBoxAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected
                ListViewSubItemAccessibleObject_GetPropertyValue_returns_correct_values
		ListViewAccessibleObject_GetPropertyValue_returns_correct_values
}
<!-- We are in TELL-MODE the following section must be completed -->

## Customer Impact

- Tests based on the old implementation of AutomationId could become obsolete.


## Regression? 

- No

## Risk

- Minimal

<!-- end TELL-MODE -->


## Test methodology <!-- How did you ensure quality? -->

- Unit tests
- CTI testing

## Test environment(s) <!-- Remove any that don't apply -->

Microsoft Windows [Version 120.2212.4170.0]
.NET Core SDK: 7.0.100-preview.3.22179.4


<!-- Mention language, UI scaling, or anything else that might be relevant -->


###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/7206)

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ButtonBase.ButtonBaseAccessibleObjectTests.cs:
```diff
@@ -235,6 +235,17 @@ public void ButtonBaseAccessibleObject_GetPropertyValue_ControlType_IsExpected_F
+        [WinFormsFact]
+        public void ButtonBaseAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected()
+        {
+            using var ownerControl = new SubButtonBase();
```
The codebase was started before `new()` syntax was available, however now this is the preferred way when writing the new code:
```suggestion
            using SubButtonBase ownerControl = new();
```

dkazennov replied:
All suggestions changed according to the preferred code style.

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs:
```diff
@@ -1555,6 +1555,17 @@ public void ControlAccessibleObject_FragmentNavigate_PreviousSibling_IsPreviousI
+        [WinFormsFact]
+        public void ControlAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected()
+        {
+            using Control ownerControl = new Control();
```
```suggestion
            using Control ownerControl = new();
```

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/GroupBoxAccessibleObjectTests.cs:
```diff
@@ -119,5 +119,16 @@ public void GroupBoxAccessibleObject_GetPropertyValue_ControlType_IsExpected_For
+        [WinFormsFact]
+        public void GroupBoxAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected()
+        {
+            using GroupBox ownerControl = new GroupBox();
```
```suggestion
            using GroupBox ownerControl = new();
```

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/PictureBox.PictureBoxAccessibleObjectTests.cs:
```diff
@@ -107,7 +107,6 @@ public void PictureBoxAccessibleObject_GetPropertyValue_Invoke_ReturnsExpected(i
             object value = pictureBoxAccessibleObject.GetPropertyValue((UIA)propertyID);
 
             Assert.Equal(expected, value);
-            Assert.False(pictureBox.IsHandleCreated);
```
Why was this check deleted? As a general rules accessing an AO must never force a owner control handle creation, and we want to ensure this rule is never broken.

dkazennov replied:
> Why was this check deleted? As a general rules accessing an AO must never force a owner control handle creation, and we want to ensure this rule is never broken.

Assert.False(pictureBox.IsHandleCreated); was being put before var pictureBoxAccessibleObject = new PictureBox.PictureBoxAccessibleObject(pictureBox); (line 105) Should it be put at the end of the test?

RussKie commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/TrackBar.TrackBarAccessibleObjectTests.cs:
```diff
@@ -413,18 +413,22 @@ public void TrackBarAccessibilityObject_GetPropertyValue_RuntimeId_ReturnsExpect
 
-            Assert.Equal(name, actual);
             Assert.False(trackBar.IsHandleCreated);
+            TrackBar.TrackBarAccessibleObject accessibilityObject = new TrackBar.TrackBarAccessibleObject(trackBar);
```
```suggestion
            TrackBar.TrackBarAccessibleObject accessibilityObject = new(trackBar);
```

RussKie reviewed (commented):
LGTM

Tanya-Solyanik commented:
>Tests based on the old implementation of AutomationId could become obsolete

Please describe what inspect and AccessibilityInsights show in cases when we return null.

Tanya-Solyanik commented:
@vladimir-krestov, @dmitrii-drobotov  - please review this change

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/LabelAccessibleObjectTests.cs:
```diff
@@ -9,20 +9,23 @@ namespace System.Windows.Forms.Tests
-        public void LabelAccessibleObject_GetPropertyValue_Name_ReturnsExpected()
+        [WinFormsTheory]
+        [InlineData((int)Interop.UiaCore.UIA.NamePropertyId, "Address")]
+        [InlineData((int)Interop.UiaCore.UIA.AutomationIdPropertyId, "Label1")]
```
Nice!

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListViewItem.ListViewItemBaseAccessibleObjectTests.cs:
```diff
@@ -336,5 +336,18 @@ public void ListViewItemBaseAccessibleObject_Toggle_WorksExpected(bool isChecked
+        [WinFormsFact]
+        public void ListViewItemBaseAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected()
+        {
+            using ListView control = new();
```
Consider `control` -> `listView`

dkazennov commented:
> > Tests based on the old implementation of AutomationId could become obsolete
> 
> Please describe what inspect and AccessibilityInsights show in cases when we return null for automation ID. This fix changes behavior of ControlAccessibleObject. Previously it was returning a null ID and now it returns Name. I wonder how this is manifested in accessibility tools.

A description was being put into PR text description.
"If AutomationId is null then Inspect will display no AutomationId property at all. It shows the property and the value if the field is overridden and is not null (a normal case)."
I'll check this with AccessibilityInsights as well.

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildEditUiaProvider.cs:
```diff
@@ -15,6 +15,7 @@ public partial class ComboBox
         internal class ComboBoxChildEditUiaProvider : ChildAccessibleObject
         {
             private const string COMBO_BOX_EDIT_AUTOMATION_ID = "1001";
+            private protected override string AutomationId => COMBO_BOX_EDIT_AUTOMATION_ID;
```
We try to place properties after the constructor, and constructor after the field definitions. Could you please move this property down?

dkazennov replied:
Fixed.

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/TabPage.TabAccessibleObject.cs:
```diff
@@ -12,6 +12,7 @@ public partial class TabPage
     {
         internal class TabAccessibleObject : AccessibleObject
         {
+            private protected override string AutomationId => _owningTabPage.Name;
```
nit: I would put AutomationId after the constructor, and owner too

dkazennov replied:
Fixed.

Tanya-Solyanik commented on src/System.Windows.Forms/src/System/Windows/Forms/PropertyGridInternal/GridEntry.GridEntryAccessibleObject.cs:
```diff
@@ -12,6 +12,8 @@ internal abstract partial class GridEntry
     {
         public class GridEntryAccessibleObject : AccessibleObject
         {
+            private protected override string AutomationId => GetHashCode().ToString();
```
nit: move to after the constructor

dkazennov replied:
Fixed.

Tanya-Solyanik reviewed (commented):
Looks good!. I added some nits and am waiting for a sign off from @dmitrii-drobotov

dkazennov commented:
> * You can also remove `UiaCore.UIA.AutomationIdPropertyId => Owner.Name` in `PrintPreviewControlAccessibleObject` and `SplitContainerAccessibleObject`

These classes introduced after this branch were created (next time I will use cherry-pick to take changes of an old commit to a new branch created from most recent main). I propose we will create a new issue regarding these two classes?

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/AccessibleObject.cs:
```diff
@@ -391,6 +391,8 @@ internal virtual int ProviderOptions
 
         internal virtual UiaCore.IRawElementProviderSimple? HostRawElementProvider => null;
 
+        private protected virtual string? AutomationId => null;
```
Please move it to be in alphabet order

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ComboBoxChildEditUiaProvider.cs:
```diff
@@ -33,6 +31,9 @@ public ComboBoxChildEditUiaProvider(ComboBox owner, IntPtr childEditControlhandl
                 UseTextProviders(_textProvider, _textProvider);
             }
 
+            private const string COMBO_BOX_EDIT_AUTOMATION_ID = "1001";
```
I would prefer to keep it as before, because it's a private const, that should be in the beginning of the class

RussKie replied:
https://docs.microsoft.com/en-us/archive/blogs/brada/internal-coding-guidelines#28-file-organization

vladimir-krestov commented on src/System.Windows.Forms/src/System/Windows/Forms/TabPage.TabAccessibleObject.cs:
```diff
@@ -12,15 +12,16 @@ public partial class TabPage
             }
 
+            private protected override string AutomationId => _owningTabPage.Name;
+            private TabControl? OwningTabControl => _owningTabPage.ParentInternal as TabControl;
```
Please move `OwningTabControl` to be in alphabet order

vladimir-krestov commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/AccessibleObjectTests.cs:
```diff
@@ -2568,12 +2568,14 @@ public unsafe void TestAccessibleObjectFromPoint_ComboBox()
+        [WinFormsTheory]
+        [InlineData((int)UiaCore.UIA.ControlTypePropertyId)]
+        [InlineData((int)UiaCore.UIA.AutomationIdPropertyId)]
+        public void AccessibleObject_GetPropertyValue_ReturnsNull(int propertyId)
```
I don't like this name, because `GetPropertyValue` returns `null` not for all cases. You can rename like that, for example.
```suggestion
        public void AccessibleObject_GetPropertyValue_ReturnsNull_ForSeveralProperties(int propertyId)
```

dkazennov replied:
I believe that AccessibleObject_GetPropertyValue_ReturnsNull_IfExpected should be good.

vladimir-krestov commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ButtonBase.ButtonBaseAccessibleObjectTests.cs:
```diff
@@ -235,6 +235,17 @@ public void ButtonBaseAccessibleObject_GetPropertyValue_ControlType_IsExpected_F
         }
 
+        [WinFormsFact]
+        public void ButtonBaseAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected()
+        {
+            using SubButtonBase ownerControl = new();
+            ButtonBase.ButtonBaseAccessibleObject accessibleObject = new(ownerControl);
+            string expected = accessibleObject.Owner.Name;
+
+            Assert.Equal(accessibleObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId), expected);
+            Assert.False(ownerControl.IsHandleCreated);
```
```suggestion
        public void ButtonBaseAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected()
        {
            using SubButtonBase ownerControl = new() {Name = "test name"};
            string expected = ownerControl.Name;
            object actual = ownerControl.AccessibleObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId);
            
            Assert.Equal(expected, actual);
            Assert.False(ownerControl.IsHandleCreated);
```

vladimir-krestov commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs:
```diff
@@ -1555,6 +1555,17 @@ public void ControlAccessibleObject_FragmentNavigate_PreviousSibling_IsPreviousI
         }
 
+        [WinFormsFact]
+        public void ControlAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected()
+        {
+            using Control ownerControl = new();
+            Control.ControlAccessibleObject accessibleObject = new Control.ControlAccessibleObject(ownerControl);
+            string expected = accessibleObject.Owner.Name;
+
+            Assert.Equal(accessibleObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId), expected);
+            Assert.False(ownerControl.IsHandleCreated);
+        }
```
```suggestion
        public void ControlAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected()
        {
            using Control ownerControl = new() {Name = "test name"};
            string expected = ownerControl.Name;
            object actual = ownerControl.AccessibleObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId);
            
            Assert.Equal(expected, actual);
            Assert.False(ownerControl.IsHandleCreated);
        }
```

dkazennov replied:
Fixed in ControlAccessibleObjectTest, ButtonBaseAccessibleObject and GroupBoxAccessibleObject.

vladimir-krestov commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DataGridViewCellAccessibleObjectTests.cs:
```diff
@@ -995,6 +995,24 @@ public void DataGridViewCellAccessibleObject_Column_ReturnExpected_IfCustomOrder
+                expected += runtimeIdPart.ToString();
+            }
+
+            Assert.Equal(dataGridViewCellAccessibleObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId), expected);
```
Replace these arguments, `expected` should be first (check the Equal method inside)

vladimir-krestov commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/GroupBoxAccessibleObjectTests.cs:
```diff
@@ -119,5 +119,16 @@ public void GroupBoxAccessibleObject_GetPropertyValue_ControlType_IsExpected_For
         }
+
+        [WinFormsFact]
+        public void GroupBoxAccessibleObject_GetPropertyValue_AutomationId_ReturnsExpected()
```
Rework this test and next like I proposed before as well

vladimir-krestov reviewed (commented):
Looks great

vladimir-krestov commented:
> > > Tests based on the old implementation of AutomationId could become obsolete
> > 
> > 
> > Please describe what inspect and AccessibilityInsights show in cases when we return null for automation ID. This fix changes behavior of ControlAccessibleObject. Previously it was returning a null ID and now it returns Name. I wonder how this is manifested in accessibility tools.
> 
> @Tanya-Solyanik A description was being put into PR text description. "If AutomationId is null then Inspect will display no AutomationId property at all. It shows the property and the value if the field is overridden and is not null (a normal case)." I'll check this with AccessibilityInsights as well.

I'm interesting, what happens with the rest controls, that have "string.Empty" now for `AutomationId`?
What AI says?

dkazennov commented:
This commit was being rebased on the most recent main. Conflicts were being resolved.

dmitrii-drobotov reviewed (dismissed):
LGTM👍

dkazennov commented:
> > > > Tests based on the old implementation of AutomationId could become obsolete
> > > 
> > > 
> > > Please describe what inspect and AccessibilityInsights show in cases when we return null for automation ID. This fix changes behavior of ControlAccessibleObject. Previously it was returning a null ID and now it returns Name. I wonder how this is manifested in accessibility tools.
> > 
> > 
> > @Tanya-Solyanik A description was being put into PR text description. "If AutomationId is null then Inspect will display no AutomationId property at all. It shows the property and the value if the field is overridden and is not null (a normal case)." I'll check this with AccessibilityInsights as well.
> 
> I'm interesting, what happens with the rest controls, that have "string.Empty" now for `AutomationId`? What AI says?

If AutomationId is null is is not being shown by default in AccessibilityInsights. If shown it has description "Property Doesn't exist". Same result if AutomationId is string.Empty.

![image](https://user-images.githubusercontent.com/102961955/170730401-3fc02afd-62de-43f9-b3d5-3b51515038c7.png)

RussKie reviewed (commented):
👍

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ButtonBase.ButtonBaseAccessibleObjectTests.cs:
```diff
@@ -235,6 +235,17 @@ public void ButtonBaseAccessibleObject_GetPropertyValue_ControlType_IsExpected_F
+            string expected = ownerControl.Name;
+            object actual = ownerControl.AccessibilityObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId);
+
+            Assert.Equal(actual, expected);
```
If you add parameter name, this would look `Assert.Equal(expected: actual, actual: expected); `  Please swap parameters.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs:
```diff
@@ -1568,6 +1568,17 @@ public void ControlAccessibleObject_FragmentNavigate_PreviousSibling_IsPreviousI
+            string expected = ownerControl.Name;
+            object actual = ownerControl.AccessibilityObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId);
+
+            Assert.Equal(actual, expected);
```
Please swap parameters.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/DataGridViewCellAccessibleObjectTests.cs:
```diff
@@ -995,6 +995,24 @@ public void DataGridViewCellAccessibleObject_Column_ReturnExpected_IfCustomOrder
+                expected += runtimeIdPart.ToString();
+            }
+
+            Assert.Equal(expected, dataGridViewCellAccessibleObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId));
```
Please swap parameters

dkazennov replied:
> Please swap parameters
Is in incorrect? `expected` comes as a first argument, `actual` as a second one.

Tanya-Solyanik replied:
You are right!

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/GroupBoxAccessibleObjectTests.cs:
```diff
@@ -119,5 +119,16 @@ public void GroupBoxAccessibleObject_GetPropertyValue_ControlType_IsExpected_For
+            string expected = ownerControl.Name;
+            object actual = ownerControl.AccessibilityObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId);
+
+            Assert.Equal(actual, expected);
```
Please swap parameters

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListViewItem.ListViewItemBaseAccessibleObjectTests.cs:
```diff
@@ -336,5 +336,18 @@ public void ListViewItemBaseAccessibleObject_Toggle_WorksExpected(bool isChecked
+            ListViewItemBaseAccessibleObject accessibleObject = (ListViewItemBaseAccessibleObject)item.AccessibilityObject;
+
+            var expected = string.Format("{0}-{1}", typeof(ListViewItem).Name, accessibleObject.CurrentIndex);
+            Assert.Equal(accessibleObject.GetPropertyValue(UiaCore.UIA.AutomationIdPropertyId), expected);
```
Please swap parameters

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ListViewItem.ListViewItemBaseAccessibleObjectTests.cs:
```diff
@@ -336,5 +336,18 @@ public void ListViewItemBaseAccessibleObject_Toggle_WorksExpected(bool isChecked
+            using ListView listView = new();
+            ListViewItem item = new();
+            listView.Items.Add(item);
+            ListViewItemBaseAccessibleObject accessibleObject = (ListViewItemBaseAccessibleObject)item.AccessibilityObject;
```
It's appropriate to use var in this case because of the explicit cast.
```suggestion
            var accessibleObject = (ListViewItemBaseAccessibleObject)item.AccessibilityObject;
```

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/TabControl.TabControlAccessibleObjectTests.cs:
```diff
@@ -744,18 +744,21 @@ public void TabControlAccessibleObject_GetPropertyValue_IsEnabled_ReturnsExpecte
             };
 
-            object actual = tabControl.AccessibilityObject.GetPropertyValue(UiaCore.UIA.NamePropertyId);
+            TabControlAccessibleObject accessibilityObject = new TabControlAccessibleObject(tabControl);
```
consider using `var` or 
```
TabControlAccessibleObject accessibilityObject = new(tabControl);
```

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/TabPage.TabPageAccessibleObjectTests.cs:
```diff
@@ -479,18 +479,21 @@ public void TabPageAccessibleObject_GetPropertyValue_AccessKey_ReturnExpected(bo
             };
 
-            object actual = tabPage.AccessibilityObject.GetPropertyValue(UIA.NamePropertyId);
+            TabPageAccessibleObject accessibilityObject = new TabPageAccessibleObject(tabPage);
```
Consider using `var` or `new(tabPage);` to match other tests in this file.

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/TrackBar.TrackBarAccessibleObjectTests.cs:
```diff
@@ -413,17 +413,21 @@ public void TrackBarAccessibilityObject_GetPropertyValue_RuntimeId_ReturnsExpect
             using TrackBar trackBar = new()
             {
-                AccessibleName = name
+                Name = "TrackBar1",
```
Consider:
```
Name = "TestControlName",
AccessibleName  = "TestAccessibleName"
```

Tanya-Solyanik reviewed (approved):
I added some stylistic nits.

Tanya-Solyanik commented:
>If AutomationId is null is is not being shown by default in AccessibilityInsights. If shown it has description "Property Doesn't exist". Same result if AutomationId is string.Empty.

@dkazennov  - Could you please double-check this result? I see that  `null` results in Automation ID property being looked up by the native Accessible object, which results in the window text most commonly. Empty string results in "property not defined". So the change is that Inheritors of ControlsAccessibleObjects that were not providing a custom AutomationId property, before the change were seeing ID provided by the MSAA proxy and after the change they will see `Owner.Name`

dkazennov commented:
> > If AutomationId is null is is not being shown by default in AccessibilityInsights. If shown it has description "Property Doesn't exist". Same result if AutomationId is string.Empty.
> 
> @dkazennov - Could you please double-check this result? I see that `null` results in Automation ID property being looked up by the native Accessible object, which results in the window text most commonly. Empty string results in "property not defined". So the change is that Inheritors of ControlsAccessibleObjects that were not providing a custom AutomationId property, before the change were seeing ID provided by the MSAA proxy and after the change they will see `Owner.Name`

**Case 1. class ListViewItemBaseAccessibleObject : AccessibleObject**

I've tested ListViewItemBaseAccessibleObject with AccessibilityInsight. ListViewItemBaseAccessibleObject has base class AccessibleObject. 
AccessibleObject has property `private protected virtual string? AutomationId => null;`
ListViewItemBaseAccessibleObject has porperty `private protected override string AutomationId => string.Format("{0}-{1}", typeof(ListViewItem).Name, CurrentIndex);`

If I change `ListViewItemBaseAccessibleObject.AutomationId` property to return **null**, AccessibilityInsight will show `Property does not exist`
If I change `ListViewItemBaseAccessibleObject.AutomationId` property to return **string.Empty**, AccessibilityInsight will show `Property does not exist`
If I change `ListViewItemBaseAccessibleObject.AutomationId` property to return **"TEST VALUE STRING"**, AccessibilityInsight will show `"TEST VALUE STRING"`

![image](https://user-images.githubusercontent.com/102961955/171427953-f709101f-0ace-4626-9a1d-e0aad255d7c5.png)

**Case 2. class ButtonBaseAccessibleObject : ControlAccessibleObject**

ControlAccessibleObject has property `private protected override string AutomationId => Owner.Name;`
ButtonBaseAccessibleObject has no `AutomationId` override property.

If tested without override property AutomationId	, AutomationId will show `Property does not exist` (probably because in WinformsConrolsTest `ControlAccessibleObject.Owner.Name` is string.Empty
If tested with override property `Button.AutomationId` = "TEST VALUE STRING", AutomationId will show "TEST VALUE STRING"
If I change ControlAccessibleObject property to `private protected override string AutomationId => "TEST VALUE STRING";` AutomationId shows "TEST VALUE STRING".

![image](https://user-images.githubusercontent.com/102961955/171432443-ec55b24a-8436-4e9b-99fa-10b126d34444.png)

So classes with base class `ControlAccessibleObject` and without overridden property `AutomationId`  will return ControlAccessibleObject.AutomationId. Which is Owner.Name (wherever it is string.Empty or some value);

null or string.Empty returns  `Property does not exist`

Tanya-Solyanik commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/TabPage.TabPageAccessibleObjectTests.cs:
```diff
@@ -479,18 +479,21 @@ public void TabPageAccessibleObject_GetPropertyValue_AccessKey_ReturnExpected(bo
             };
 
-            object actual = tabPage.AccessibilityObject.GetPropertyValue(UIA.NamePropertyId);
+            TabPageAccessibleObject accessibilityObject = new(tabPage);
```
It will be close to real live scenario to get Accessible object from the control rather than new up one.

Tanya-Solyanik reviewed (commented):
Added some cleanup-y comments

dkazennov commented:
> > null or string.Empty returns Property does not exist
> 
> Had you rebuilt the winforms dll properly when testing this?

I've rebuild the solution, commented out the overridden `AutomationId` property in `ControlAccessibleObject` (`ControlAccessibleObject.AutomationId` returns Owner.Name and it's base `AccessibleObject.AutomationId` returns null) and started the WinformsControlsTest. I checked button control (it has no overridden property and therefore returns null) and AutomationId in both Ispector and AccessibiltyInsights was a number as shown on a screenshot.

![image](https://user-images.githubusercontent.com/102961955/171574240-ab496f3a-d70c-4108-910a-6fd441049067.png)

dkazennov commented on src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/LabelAccessibleObjectTests.cs:
```diff
@@ -9,20 +9,23 @@ namespace System.Windows.Forms.Tests
+            };
 
-            Assert.False(label.IsHandleCreated);
+            Label.LabelAccessibleObject accessibilityObject = new Label.LabelAccessibleObject(label);
```
> Why are you not getting Accessible object from your `label` under test? `label.AccessibleObject`?

Oh, I think I've missed it since the last implementation of the test.

Tanya-Solyanik replied:
>Btw, all existing tests in LabelAccessibleObjectTests.cs get AccessibleObject thought the constructor LabelAccessibleObject(label); Should it be changed as well?

Great find! yes please do. Feel free to do it either in this or the next PR.

dkazennov replied:
> Great find! yes please do. Feel free to do it either in this or the next PR.
Since there is a CTI testing already requested I believe we do this in next PR.

Tanya-Solyanik commented:
> I checked button control (it has no overridden property and therefore returns null) and AutomationId in both Ispector and AccessibiltyInsights was a number as shown on a screenshot.

Thank you for the experiment. This result matches the pattern that when managed accessible object return null for aq property, UIA reads the value from the default implementation in MSAA proxy.

dkazennov commented:
> > I checked button control (it has no overridden property and therefore returns null) and AutomationId in both Ispector and AccessibiltyInsights was a number as shown on a screenshot.
> 
> Thank you for the experiment. This result matches the pattern that when managed accessible object return null for aq property, UIA reads the value from the default implementation in MSAA proxy.

@Tanya-Solyanik 
Should there be any changes in this fix then? Or does only concerns documentation?

Tanya-Solyanik commented:
>Should there be any changes in this fix then? Or does only concerns documentation?

I think this is an acceptable break as long as it is documented.

RussKie reviewed (commented):
👍

RussKie commented:
@Tanya-Solyanik I see you applied "Documentation breaking" label. Could you please elaborate what's breaking here?

dkazennov commented:
@Tanya-Solyanik
CTI testing confirmed our results regarding  the pattern that when managed accessible object return null for aq property, UIA reads the value from the default implementation in MSAA proxy.

Tanya-Solyanik commented:
> Could you please elaborate what's breaking here?

We are returning different value for AutomationID property. Exactly how different, is not clear as we are getting conflicting results.

Tanya-Solyanik commented:
@dkazennov - CTI test results are not quite what I expected. I'd like to understand why ComboBox.Name=null results in AutomationID as "property not available" instead of a numeric value.

dkazennov commented:
> @dkazennov - CTI test results are not quite what I expected. I'd like to understand why ComboBox.Name=null results in AutomationID as "property not available" instead of a numeric value.

I see. I'll try to figure it out than.

RussKie commented:
> Winforms code does not expose this property to be easily overridden directly.

Ah, yes, it vaguely rings true for me. We had to subclass a type and its AccessibleObject to override the accessibility values: https://github.com/gitextensions/gitextensions/pull/9887#issuecomment-1063995067

dkazennov commented:
@Tanya-Solyanik 
I have found that in WinformsControlsTests ComboBoxAccessibleObject (inherits ControlAccessibleObject) and ButtonBaseAccessibleObject (inherits ControlAccessibleObject) behave in the same way.

1. If the property `private protected override string AutomationId => Owner.Name;` is present in `ControlAccessibleObject` and controls have no Names (commented out `this.comboBox2.Name = "comboBox2"` in `ComboBoxes.Designer.cs`) then Insights will show ComboBoxAccessibleObject.AutomationId and ButtonBaseAccessibleObject.AutomationId as 'Property does not exist'.

2.  If the property `private protected override string AutomationId => Owner.Name;` is commented out in `ControlAccessibleObject` and therefore if `private protected virtual string? AutomationId => null` in `AccessibleObject` class plays its role, then Insights will show ComboBoxAccessibleObject.AutomationId and ButtonBaseAccessibleObject.AutomationId as numeric values based on hashcode.

3.  If the property `private protected override string AutomationId => Owner.Name;` is replaced with `private protected override string AutomationId => null;` in `ControlAccessibleObject` then Insights will show ComboBoxAccessibleObject.AutomationId and ButtonBaseAccessibleObject.AutomationId as numeric values based on hashcode.

4. If Owner.Name is set to null or string.Empty in `ControlAccessibleObject.` constructorthen Insights will show ComboBoxAccessibleObject.AutomationId and ButtonBaseAccessibleObject.AutomationId as 'Property does not exist'.

So the exact source of the issue is the new property `private protected override string AutomationId` in `ControlAccessibleObject`. If the property returns null, it is being show as a numeric value based of hash.

![image](https://user-images.githubusercontent.com/102961955/172446203-b963f114-5de4-4a60-8838-8c7e4d0580fa.png)

RussKie commented:
> 2\. If the property `private protected override string AutomationId => Owner.Name;` is commented out in `ControlAccessibleObject` and therefore if `private protected virtual string? AutomationId => null` in `AccessibleObject` class plays its role, then Insights will show ComboBoxAccessibleObject.AutomationId and ButtonBaseAccessibleObject.AutomationId as numeric values based on hashcode.

Do we have tests for this?

dkazennov commented:
> > 2. If the property `private protected override string AutomationId => Owner.Name;` is commented out in `ControlAccessibleObject` and therefore if `private protected virtual string? AutomationId => null` in `AccessibleObject` class plays its role, then Insights will show ComboBoxAccessibleObject.AutomationId and ButtonBaseAccessibleObject.AutomationId as numeric values based on hashcode.
> 
> Do we have tests for this?

No, we don't. Is it something to be tested with a unit test?

dkazennov commented:
**Conclusion:**

If the property `AutomationId` returns `null` (but not `Owner.Name` even if `Owner.Name` is null or empty) then Accessibility Tools will show AutomationId (e.g. ComboBoxAccessibleObject.AutomationId and ButtonBaseAccessibleObject.AutomationId) as numeric values based on a hashcode.

So on the Windows Forms layer the value is null (in all concluded unit tests Assert.Null(actual AutomationId) would be true ), but on the MSAA Proxy level (between MSAA Proxy and Accessibility Tools) this value is being replaced by a native hashcode.

dkazennov commented:
Rebased on the most recent main branch. 
Implemented changes to WebBrowser.WebBrowserAccessibleObject from recent PR https://github.com/dotnet/winforms/pull/7314

Tanya-Solyanik reviewed (approved):
Thank you!
